go/src/github.com/mesosphere/rendler/scheduler/main.go

```diff
@@ -1,281 +1,284 @@
+	return task
+}
 
-	driver := mesos.SchedulerDriver{
-		Master: *master,
-		Framework: mesos.FrameworkInfo{
-			Name: proto.String("RENDLER"),
-			User: proto.String(""),
-		},
+func (s *RendlerScheduler) newRenderTask(url string, offer *mesos.Offer) *mesos.TaskInfo {
+	task := s.newTaskPrototype(offer)
+	task.Name = proto.String("RENDER_" + *task.TaskId.Value)
+	task.Executor = s.renderExecutor
+	task.Data = []byte(url)
+	return task
+}
 
-		Scheduler: &mesos.Scheduler{
+func (s *RendlerScheduler) Registered(
+	driver sched.SchedulerDriver,
+	frameworkId *mesos.FrameworkID,
+	masterInfo *mesos.MasterInfo) {
+	log.Printf("Framework %s registered with master %s", frameworkId, masterInfo)
+}
 
-			Registered: func(
-				driver *mesos.SchedulerDriver,
-				frameworkId mesos.FrameworkID,
-				masterInfo mesos.MasterInfo) {
-				log.Printf("Registered")
-			},
+func (s *RendlerScheduler) Reregistered(driver sched.SchedulerDriver, masterInfo *mesos.MasterInfo) {
+	log.Printf("Framework re-registered with master %s", masterInfo)
+}
 
-			ResourceOffers: func(driver *mesos.SchedulerDriver, offers []mesos.Offer) {
-				printQueueStatistics()
-
-				for _, offer := range offers {
-					if shuttingDown {
-						fmt.Println("Shutting down: declining offer on [", offer.Hostname, "]")
-						driver.DeclineOffer(offer.Id)
-						continue
-					}
-
-					tasks := []mesos.TaskInfo{}
-
-					for i := 0; i < maxTasksForOffer(offer)/2; i++ {
-						if crawlQueue.Front() != nil {
-							url := crawlQueue.Front().Value.(string)
-							crawlQueue.Remove(crawlQueue.Front())
-							task := makeCrawlTask(url, offer)
-							tasks = append(tasks, *task)
-						}
-						if renderQueue.Front() != nil {
-							url := renderQueue.Front().Value.(string)
-							renderQueue.Remove(renderQueue.Front())
-							task := makeRenderTask(url, offer)
-							tasks = append(tasks, *task)
-						}
-					}
-
-					if len(tasks) == 0 {
-						driver.DeclineOffer(offer.Id)
-					} else {
-						driver.LaunchTasks(offer.Id, tasks)
-					}
-				}
-			},
+func (s *RendlerScheduler) Disconnected(sched.SchedulerDriver) {
+	log.Println("Framework disconnected with master")
+}
 
-			StatusUpdate: func(driver *mesos.SchedulerDriver, status mesos.TaskStatus) {
-				log.Printf("Received task status [%s] for task [%s]", rendler.NameFor(status.State), *status.TaskId.Value)
+func (s *RendlerScheduler) ResourceOffers(driver sched.SchedulerDriver, offers []*mesos.Offer) {
+	log.Printf("Received %d resource offers", len(offers))
+	for _, offer := range offers {
+		select {
+		case <-s.shutdown:
+			log.Println("Shutting down: declining offer on [", offer.Hostname, "]")
+			driver.DeclineOffer(offer.Id, DefaultFilter)
+			continue
+		default:
+		}
 
-				if *status.State == mesos.TaskState_TASK_RUNNING {
-					tasksRunning++
-				} else if rendler.IsTerminal(status.State) {
-					tasksRunning--
-				}
-			},
+		tasks := []*mesos.TaskInfo{}
+		// += 2 because we create 2 tasks for each iteration
```

Comment:
This isn't necessarily true. If the front of any of the queues is nil, we shouldn't increment by 2.